File: python/containerwithmostwater.py
from typing import List


class Solution:
    def maxArea(self, heights: List[int]) -> int:
        max_vol = 0
        left, right = 0, len(heights) - 1
        while left < right:
            # while heights[left] == heights[left + 1]:
            #     left += 1
            # while heights[right] == heights[right - 1]:
            #     right -= 1

            base = right - left
            min_height = min(heights[left], heights[right])
            volumn = base * min_height

            max_vol = max(max_vol, volumn)
            if heights[right] < heights[left]:
                right -= 1
            else:
                left += 1
        return max_vol


sol = Solution()
print(sol.maxArea([1, 7, 2, 5, 4, 7, 3, 6]))

File: c++/containerwithmostwater.cpp
class Solution {
public:
  int maxArea(vector<int> &heights) {
    // idea container depend on max(min(left, right)*(right - left))
    int left = 0;
    int right = heights.size() - 1;
    int res = 0;
    while (left < right) {
      int height = min(heights[left], heights[right]);
      int base = right - left;
      res = max(res, base * height);
      if (heights[left] <= heights[right]) {
        left++;
      } else {
        right--;
      }
    }
    return res;
  }
};
